File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        string s_prime = "#";
        for (char c : s) {
            s_prime += c;
            s_prime += "#";
        }

        int n = s_prime.length();
        vector<int> palindrome_radii(n, 0);
        int center = 0;
        int radius = 0;

        for (int i = 0; i < n; i++) {
            int mirror = 2 * center - i;

            if (i < radius) {
                palindrome_radii[i] = min(radius - i, palindrome_radii[mirror]);
            }

            while (i + 1 + palindrome_radii[i] < n &&
                   i - 1 - palindrome_radii[i] >= 0 &&
                   s_prime[i + 1 + palindrome_radii[i]] ==
                       s_prime[i - 1 - palindrome_radii[i]]) {
                palindrome_radii[i]++;
            }

            if (i + palindrome_radii[i] > radius) {
                center = i;
                radius = i + palindrome_radii[i];
            }
        }

        int max_length = 0;
        int center_index = 0;
        for (int i = 0; i < n; i++) {
            if (palindrome_radii[i] > max_length) {
                max_length = palindrome_radii[i];
                center_index = i;
            }
        }

        int start_index = (center_index - max_length) / 2;
        string longest_palindrome = s.substr(start_index, max_length);

        return longest_palindrome;
    }
};


/*
class Solution {
public:

    int isPalindrome(string& s, int left, int right){
        while(left >= 0 && right < s.size() && s[left] == s[right]){
            left--;
            right++;
        }
        return right-left-1;
    }
    string longestPalindrome(string s) {
        int n = s.size();
        if (s.empty())
		    return "";

        // Initializing start and end indexes of resultant substring as zero
	    int start = 0, end = 0;

        for(int i = 0; i < n; i++){
            int len1 = isPalindrome(s,i,i); // try for odd length
            int len2 = isPalindrome(s,i,i+1); // try for even length

            //Now we have both half length len1 and len2 so we now calculate the max length of both
            int len = max(len1,len2);
            
            if(len > end-start){
                // Updating start and end indexes of palindromic substring
			start = i - (len - 1) / 2;
			end = i + len / 2;
            }
        }

        return s.substr(start, end+1-start);
    }
};
*/

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    // Helper function to find the length of the longest palindrome centered at 'leftCenter' and 'rightCenter'
    public int expandAroundCenter(String inputString, int leftCenter, int rightCenter) {
        // Expand outwards as long as the characters match and within bounds
        while (leftCenter >= 0 && rightCenter < inputString.length() && 
               inputString.charAt(leftCenter) == inputString.charAt(rightCenter)) {
            leftCenter--;  // Move left pointer leftward
            rightCenter++; // Move right pointer rightward
        }

        // Return the length of the found palindrome
        return rightCenter - leftCenter - 1;
    }

    public String longestPalindrome(String inputString) {
        int stringLength = inputString.length(); // Get the length of the input string

        if (stringLength == 0) {
            return ""; // Edge case: if the string is empty, return an empty string
        }

        int longestPalindromeStart = 0; // Stores the starting index of the longest palindrome
        int longestPalindromeEnd = 0;   // Stores the ending index of the longest palindrome

        // Iterate through each character in the string
        for (int currentIndex = 0; currentIndex < stringLength; currentIndex++) {
            // Check for odd-length palindrome (single character center)
            int oddLengthPalindrome = expandAroundCenter(inputString, currentIndex, currentIndex);

            // Check for even-length palindrome (double character center)
            int evenLengthPalindrome = expandAroundCenter(inputString, currentIndex, currentIndex + 1);

            // Get the maximum length palindrome from both cases
            int maxPalindromeLength = Math.max(oddLengthPalindrome, evenLengthPalindrome);

            // If the current palindrome is longer than the previous longest
            if (maxPalindromeLength > longestPalindromeEnd - longestPalindromeStart) {
                // Update start and end indices based on the center position
                longestPalindromeStart = currentIndex - (maxPalindromeLength - 1) / 2;
                longestPalindromeEnd = currentIndex + maxPalindromeLength / 2;
            }
        }

        // Extract and return the longest palindrome substring
        return inputString.substring(longestPalindromeStart, longestPalindromeEnd + 1);
    }
}


/*
class Solution {
    // Helper function to check the longest palindrome centered at 'left' and 'right'
    public int isPalindrome(String s, int left, int right) {
        // Expand outwards as long as the characters match and within bounds
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;  // Move left pointer leftward
            right++; // Move right pointer rightward
        }

        return right - left - 1; // Return the length of the found palindrome
    }

    public String longestPalindrome(String s) {
        int n = s.length(); // Get the length of the string

        if (n == 0) {
            return ""; // Edge case: if the string is empty, return an empty string
        }

        int start = 0; // Stores the starting index of the longest palindrome
        int end = 0;   // Stores the ending index of the longest palindrome

        // Iterate through each character in the string
        for (int i = 0; i < n; i++) {
            int l1 = isPalindrome(s, i, i);      // Check for odd-length palindrome (single character center)
            int l2 = isPalindrome(s, i, i + 1);  // Check for even-length palindrome (double character center)

            int maxLen = Math.max(l1, l2); // Get the maximum length palindrome from both cases

            // If the current palindrome is longer than the previous longest
            if (maxLen > end - start) {
                start = i - (maxLen - 1) / 2; // Update start index based on the center position
                end = i + maxLen / 2;         // Update end index based on the center position
            }
        }

        return s.substring(start, end + 1); // Extract the longest palindrome substring
    }
}

*/